terragraph_planner/data/template.yaml

```diff
@@ -1,106 +1,104 @@
-# This file is a template for the config yaml file with the default values used
-# in the planner, except all the file paths and DEVICE_LIST.
-# Change the values to customize the config of your own plan.
+# This file is a template for the configuration yaml file with the default
+# values used in the planner, except all the file paths and DEVICE_LIST. Change
+# the values to customize the configuration of your own plan.
 #
 # NOTE:
-# 1. The default value of each file path is an empty string rather than the string
-#    in this template.
-# 2. If any field is deleted, left blank or filled out with "null" or "~",
-#    the default value is used, except BOUNDARY_POLYGON_FILE_PATH and DEVICE_LIST.
-# 3. BOUNDARY_POLYGON_FILE_PATH must have valid values when los computation is ran.
-# 4. DEVICE_LIST must have valid values.
-# 5. The "[Optional]" tag for each field means it can be null and its default Python
-#    value is None. Other fields without that tag cannot be assigned as null.
-# 6. It's not recommended to leave a field blank.
-# 7. "None" is treated as a string "None" rather than None type in Python. Please
-#    refer to https://yaml.org/ for more information.
+# 1. BOUNDARY_POLYGON_FILE_PATH is a required field when LOS analysis is run.
+# 2. The DEVICE_LIST cannot be empty.
+# 3. If any field is deleted, left blank or filled out with "null" or "~", the
+#    default value is used. This does not apply to BOUNDARY_POLYGON_FILE_PATH
+#    and DEVICE_LIST.
+# 4. The default value of each file path is null rather than the string in this
+#    template.
+# 5. The "[Optional]" tag for each field means it can be null and its default
+#    Python value is None. Other fields without that tag cannot be assigned as
+#    null or omitted entirely.
```

Comment:
Remind me, if we get null for a non optional field, do we throw an error?

Reply:
No. The new rule 3 applies for every field except DEVICE_SKU and BOUNDARY_FILE_PATH, rather than only optional fields.

Reply:
We delete `None` fields from the dict when constructing the config object to make those `None` parameters use defaults:

https://github.com/terragraph/terragraph-planner/blob/main/terragraph_planner/common/configuration/configs.py#L142

Reply:
So really this should read:
The "[Optional]" tag for each field means its default value in Python is None.
It seems everything else is not correct.

Reply:
Python itself never prevents us from setting a non-optional argument as None. But for those parameters without [Optional] tag, we don't expect it to be None; otherwise the planner will fail at some point (not at the moment when calling the constructor). For example, `default_dn_height_on_pole` is not Optional. If we set it None, the planner will not throw an exception until we actually use it in function `infer_input_site_location`. That's why I said "[Optional]" means the variable can be None. Variables without that tag does not have None as default and will cause plan failing at some point.